pkg/agent/flowexporter/exporter/priorityqueue/priorityqueue.go

```diff
@@ -0,0 +1,101 @@
+// distributed under the License is distributed on an “AS IS” BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package priorityqueue
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+
+	"antrea.io/antrea/pkg/agent/flowexporter"
+)
+
+type ExpirePriorityQueue struct {
+	items             []*flowexporter.ItemToExpire
+	ActiveFlowTimeout time.Duration
+	IdleFlowTimeout   time.Duration
+	Mutex             sync.RWMutex
+}
+
+func NewExpirePriorityQueue(activeFlowTimeout time.Duration, idleFlowTimeout time.Duration) *ExpirePriorityQueue {
+	return &ExpirePriorityQueue{
+		items:             make([]*flowexporter.ItemToExpire, 0),
+		ActiveFlowTimeout: activeFlowTimeout,
+		IdleFlowTimeout:   idleFlowTimeout,
+	}
+}
+
+func (pq *ExpirePriorityQueue) Len() int {
+	return len(pq.items)
+}
+
+func (pq *ExpirePriorityQueue) MinExpireTime(i int) time.Time {
+	if pq.items[i].ActiveExpireTime.Before(pq.items[i].IdleExpireTime) {
+		return pq.items[i].ActiveExpireTime
+	} else {
+		return pq.items[i].IdleExpireTime
+	}
+}
+
+func (pq *ExpirePriorityQueue) Less(i, j int) bool {
+	return pq.MinExpireTime(i).Before(pq.MinExpireTime(j))
+}
+
+func (pq *ExpirePriorityQueue) Swap(i, j int) {
+	pq.items[i], pq.items[j] = pq.items[j], pq.items[i]
+	pq.items[i].Index = i
+	pq.items[j].Index = j
+}
+
+func (pq *ExpirePriorityQueue) Push(x interface{}) {
+	n := len((*pq).items)
+	item := x.(*flowexporter.ItemToExpire)
+	item.Index = n
+	(*pq).items = append((*pq).items, item)
+}
+
+func (pq *ExpirePriorityQueue) Pop() interface{} {
+	n := len((*pq).items)
+	item := ((*pq).items)[n-1]
+	item.Index = -1
+	(*pq).items = ((*pq).items)[0:(n - 1)]
+	return item
+}
+
+// Peek returns the item at the beginning of the queue, without removing the
+// item or otherwise mutating the queue. It is safe to call directly.
+func (pq *ExpirePriorityQueue) Peek() *flowexporter.ItemToExpire {
+	return pq.items[0]
+}
+
+// Update modifies the priority and flow record of an Item in the queue.
+func (pq *ExpirePriorityQueue) Update(item *flowexporter.ItemToExpire, connKey flowexporter.ConnectionKey, conn *flowexporter.Connection, activeExpireTime time.Time, idleExpireTime time.Time) {
+	item.ConnKey = connKey
+	item.Conn = *conn
+	item.ActiveExpireTime = activeExpireTime
+	item.IdleExpireTime = idleExpireTime
+
+	heap.Fix(pq, item.Index)
+}
+
+// ResetItemToExpire updates stats and TCPState in pqItem after it is sent successfully.
+func (pq *ExpirePriorityQueue) ResetItemToExpire(pqItem *flowexporter.ItemToExpire, currTime time.Time) {
+	pqItem.PrevPackets = pqItem.Conn.OriginalPackets
+	pqItem.PrevBytes = pqItem.Conn.OriginalBytes
+	pqItem.PrevReversePackets = pqItem.Conn.ReversePackets
+	pqItem.PrevReverseBytes = pqItem.Conn.ReverseBytes
+	pqItem.PrevTCPState = pqItem.Conn.TCPState
+	pqItem.ActiveExpireTime = currTime.Add(pq.ActiveFlowTimeout)
```

Comment:
I don't see why resetting these fields should be the responsibility of the priority queue